File: backtracking/N_Queen.cpp
#include <bits/stdc++.h>
using namespace std;

bool isused1[40];
bool isused2[40];
bool isused3[40];
int cnt = 0;

void solve(int cur, int n) {
    if(cur==n){
        cnt++;
        return;
    }

    for(int i=0;i<n;i++) {
        if (isused1[i] || isused2[i+cur] || isused3[n-1+cur-i]) continue;
        isused1[i] = 1;
        isused2[cur+i] = 1;
        isused3[n-1+cur-i] = 1;
        solve(cur+1, n);
        isused1[i] = 0;
        isused2[cur+i] = 0;
        isused3[n-1+cur-i] = 0;
    }
}

int main(void) {
    int n;
    cin>>n;
    solve(0,n);
    cout<<cnt;
}

File: backtracking/N_Queen.py
import sys
input = sys.stdin.readline

isused1 = [0]*40  # column(열)을 차지하고 있는지
isused2 = [0]*40  # / 방향 대각선을 차지하고 있는지
isused3 = [0]*40  # \ 방향 대각선을 차지하고 있는지
count = 0

def nq(cur, n):
    global count
    
    if cur == n:
        count += 1
        return
        
    for i in range(n):  # cur행, i열에 퀸을 놓을 예정
        if isused1[i] or isused2[i+cur] or isused3[n-1+cur-i]:
            continue
        isused1[i] = 1
        isused2[cur+i] = 1
        isused3[n-1+cur-i] = 1
        nq(cur+1, n)
        isused1[i] = 0
        isused2[cur+i] = 0
        isused3[n-1+cur-i] = 0

def main():
    n = int(input())
    nq(0, n)
    print(count)

if __name__ == "__main__":
    main()
